### ISSUE TITLE
Reconnection with the AgentX master does not succeed

### ISSUE THREAD

**Issue by mschneid** [acassen/keepalived#1083]
Hi

We noticed that from time to time the communication between the `SNMPd` and `keepalived` dies. We can reproduce this issue even with the latest commit 37ab724e75f129bb81e91a2e848e8d0634b8a1a5.

keepalived -v
```
Keepalived v2.0.10 (12/10,2018)

Copyright(C) 2001-2018 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 4.4.162
Running on Linux 4.4.0-140-generic #166-Ubuntu SMP Wed Nov 14 20:09:47 UTC 2018

configure options: --build=x86_64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir=${prefix}/lib/x86_64-linux-gnu --libexecdir=${prefix}/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --enable-snmp-vrrp --disable-lvs --disable-libipset --disable-fwmark --disable-libiptc --disable-routes --enable-strict-config-checks --disable-nftables build_alias=x86_64-linux-gnu CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2

Config options:  STRICT_CONFIG IPTABLES_CMD VRRP VRRP_AUTH OLD_CHKSUM_COMPAT SNMP_VRRP

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF RTA_ENCAP RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS RTA_VIA FRA_OIFNAME IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA NET_LINUX_IF_H_COLLISION LIBIPTC_LINUX_NET_IF_H_COLLISION VRRP_VMAC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SCHED_RT SCHED_RESET_ON_FORK
```

## Configuration to reproduce the  issue

/etc/snmp/snmpd.conf
```bash
agentAddress  udp:127.0.0.1:161
rocommunity public  localhost
master          agentx
```

/etc/default/keepalived
```bash
KEEPALIVED_OPTIONS="-D --snmp"
```

/etc/keepalived/keepalived.conf
```bash
global_defs {
    enable_script_security
}

vrrp_instance MPP-16 {
    state BACKUP
    interface ens33
    virtual_router_id 16
    priority 150
    advert_int 1

    virtual_ipaddress {
	192.0.2.10/24 dev ens33
    }
}
```

## Steps to reproduce the  issue

1. Start both Daemons
1. `snmpget -v2c -c public localhost .1.3.6.1.4.1.9586.100.5.2.3.1.4.1` succeeds
1. Stop SNMPd with `kill -STOP <PID>`. After that keepalived reports `AgentX master agent failed to respond to ping.  Attempting to re-register.` in syslog
1. `snmpget -v2c -c public localhost .1.3.6.1.4.1.9586.100.5.2.3.1.4.1` fails as expected
1. Continue SNMPd with `kill -CONT <PID>`. After this step, Keepalived repeats the following messages 
```bash
NET-SNMP version 5.7.3 AgentX subagent connected
AgentX master disconnected us, reconnecting in 15
NET-SNMP version 5.7.3 AgentX subagent connected
AgentX master disconnected us, reconnecting in 15
```
6. `snmpget -v2c -c public localhost .1.3.6.1.4.1.9586.100.5.2.3.1.4.1` is still failing

In contrast to the above procedure, the reconnection works if you stop/start the `SNMPd` correctly.

**Comment by pqarmitage** [comment]
Many thanks for this issue report which I have been able to reproduce.

It looks as though this issue started when we moved from using select() to epoll_wait(). The problem is that net-snmp doesn't provide a support function for using epoll. The problem appears to be that using epoll is stateful whereas using select is stateless; in other words when select is called, the file descriptors of interest are specified at the time of the system call, whereas epoll_wait() uses the file descriptors that have been registered. With epoll, if a file descriptor is closed and subsequently reopened, the interest in that file descriptor has to be re-registered, whereas with select() that is not the case.

Unfortunately net-snmp doesn't directly inform keepalived of errors, but it does call an error logging function when there are errors. It might be that in that case keepalived can re-register its interest in the snmp file descriptors. I'll look into this further.
